### PR TITLE
Refactor Operator hierarchy

### LINKED ISSUE #1538: Rework Ompizer-Accizer hierarchy
We need to revisit the hierarchy that implements the passes generating parallel CPU/GPU code. 

Currently Accizer inherits from Ompizer, but what we should rather have is an abstract class, with Ompizer and Accizer being two different subclasses.

The DeviceOmpizer class should also be a sibling of Ompizer and Accizer. Actually, what we probably want to have is something along these lines:

```
                        --------- Parallelizer ---------
                        |                              |
         ---- OmpizerBasic ----                   AccizerDevice
        |                     |
OmpizerHost            OmpizerDevice
```

### PR DESCRIPTION
**Pull request by FabioLuporini**
This is a big PR but I'm just refactoring code:

* OpenMP" and OpenACC inherit from common classes rather than OpenACC inheriting from OpenMP. 
* The concept of `Target` is introduced. A `Target` is in essence a triple `<Parallelizer, DataManager, Orchestrator>` which exposes a well-defined interface to Operators. The advantages in terms of code reuse and maintainability are huge. Just take a look at the new files in `devito/core`.
* The concepts of "language" and "transformer" are now neatly separated in `devito/passes/iet`.  This again vastly improves code maintainability and extensibility. Have a look at `devito/passes/iet/langbase.py` , `devito/passes/iet/parpragma.py`, and the various specializations inside `devito/passes/iet/languages/...` 
* The OpenACC and OpenMP `initialize` passes have been merged into an `initialize` pass in a common superclass 

This PR does **not** add any new functionalities. 

fixes #1538

### CODE REVIEW

File: devito/passes/iet/langbase.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+
+    @property
+    def Prodder(self):
+        return self.lang.Region
+
+
+class DeviceAwareMixin(object):
+
+    @iet_pass
+    def initialize(self, iet):
```
Comment (FabioLuporini) [inline devito/passes/iet/langbase.py:209]: @mloubout the merged `initialize`

**Comment by mloubout** [comment]
very nice cleanup, gtg.

File: devito/passes/iet/misc.py [hunk 2]
```diff
@@ -116,3 +117,34 @@ def relax_incr_dimensions(iet, **kwargs):
...
     return iet, {'efuncs': efuncs}
+
+
+def is_on_device(maybe_symbol, gpu_fit, only_writes=False):
+    """
+    True if all given Functions are allocated in the device memory, False otherwise.
+
+    Parameters
+    ----------
+    maybe_symbol : Indexed or Function or Node
```
Comment (georgebisbas) [inline devito/passes/iet/misc.py:128]: This var name seems a bit off to me? Something better?

**Review by georgebisbas — APPROVED** [review]
If I am not wrong, this helps to do openacc tile a step quicker @italoaug @FabioLuporini ?

`OpenMP" and OpenACC inherit from common classes rather than OpenACC inheriting from OpenMP.`

Seems ok to me. Should I test something else?

File: devito/passes/iet/misc.py (hunk 2 above)
Comment (FabioLuporini) [reply devito/passes/iet/misc.py:128]: I think it's fine, but I'm open to suggestions

File: devito/core/gpu.py [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+from devito.passes.iet import (DeviceOmpTarget, DeviceAccTarget, optimize_halospots,
+                               mpiize, hoist_prodders, is_on_device)
+from devito.tools import as_tuple, timed_pass
+
+__all__ = ['DeviceNoopOperator', 'DeviceAdvOperator', 'DeviceCustomOperator',
+           'DeviceNoopOmpOperator', 'DeviceAdvOmpOperator', 'DeviceCustomOmpOperator',
+           'DeviceNoopAccOperator', 'DeviceAdvAccOperator', 'DeviceCustomAccOperator']
+
+
+class DeviceOperatorMixin(object):
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/core/gpu.py:20]: You need to learn what [a mixin is](https://en.wikipedia.org/wiki/Mixin)

**Comment by FabioLuporini** [comment]
@georgebisbas 

> If I am not wrong, this helps to do openacc tile a step quicker 

yeah, this is massive cleanup, hopefully yes!

> Seems ok to me. Should I test something else?

No, we're good here. Thanks!

**Comment by FabioLuporini** [comment]
Approved; CI-gpu ran fine (https://github.com/devitocodes/devito/actions/runs/530166715, id 2214); hence merging

**Pull request merged**